File: sca/runtime/core/src/tuscany/sca/model/Composite.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 *   
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* $Rev$ $Date$ */

#include "tuscany/sca/util/Logging.h"
#include "tuscany/sca/util/Utils.h"
#include "tuscany/sca/model/Composite.h"
#include "tuscany/sca/model/Component.h"
#include "tuscany/sca/model/Wire.h"
#include "tuscany/sca/model/WSDLDefinition.h"
#include "tuscany/sca/model/Service.h"
#include "tuscany/sca/model/Reference.h"
#include "tuscany/sca/model/ServiceBinding.h"
#include "tuscany/sca/model/ReferenceBinding.h"

namespace tuscany
{
    namespace sca
    {
        
        namespace model
        {
            
            // Constructor
            Composite::Composite(const string& name, const string& root) 
                : ComponentType(0, name), root(root)
            {
                LOGENTRY(1, "Composite::constructor");
                LOGEXIT(1, "Composite::constructor");
            }
            
            // Destructor
            Composite::~Composite()
            {
            }
            
            void Composite::addComponent(Component* component)
            {
                LOGENTRY(1, "Composite::addComponent");
                components[component->getName()] = component;
                LOGEXIT(1, "Composite::addComponent");
            }
            
            Component* Composite::findComponent(const std::string& name)
            {
                LOGENTRY(1, "Composite::findComponent");
                Component* component = components[name];
                LOGEXIT(1, "Composite::findComponent");
                return component;
            }
            
            Service* Composite::findComponentService(const std::string& name)
            {
                LOGENTRY(1, "Composite::findComponentService");
                
                Service* service = 0;
                
                string componentName;
                string serviceName;
                Utils::tokeniseUri(name, componentName, serviceName);
                
                // Locate the component
                Component* component = findComponent(componentName);
                if (component)
                {
                    // Locate the service
                    service = component->findService(serviceName);
                }
                LOGEXIT(1, "Composite::findComponentService");
                return service;
            }
            
            void Composite::addWire(const std::string& source, const std::string& target)
            {
                LOGENTRY(1, "Composite::addWire");
                Wire* wire=new Wire(source, target);
                wires.push_back(wire);
                LOGEXIT(1, "Composite::addWire");
            }
            
            void Composite::addInclude(Composite* composite)
            {
                LOGENTRY(1, "Composite::addInclude");
                includes.push_back(composite);

                for (COMPONENT_MAP::iterator iter = composite->components.begin();
                iter != composite->components.end();
                iter++)
                {
                    components[iter->first] = iter->second;
                } 
                LOGEXIT(1, "Composite::addInclude");
            }
            
            void Composite::resolveWires()
            {
                LOGENTRY(1, "Composite::resolveWires");

                for (WIRES::iterator iter = wires.begin();
                iter != wires.end();
                iter++)
                {
                    Wire* wire = *iter;
                    
                    // Locate the target
                    Service* service = findComponentService(wire->getTarget());
                    if (!service)
                    {
                        LOGERROR_1(0, "Composite::resolveWires: Wire target %s not found", wire->getTarget().c_str());
                    }
                    else
                    {
                        Component* component = findComponent(wire->getSourceComponent());
                        if (component)
                        {
                            Reference* reference = component->findReference(wire->getSourceReference());
                            if (reference)
                            {
                                
                                // Configure the binding on the reference from the binding on the target
                                // service
                                reference->getBinding()->configure(service->getBinding());
                            }
                            else
                            {
                                LOGERROR_1(0, "Composite::resolveWires: Wire source reference %s not found", wire->getSourceReference().c_str());
                            }
                        }
                        else
                        {
                            LOGERROR_1(0, "Composite::resolveWires: Wire source %s not found", wire->getSourceComponent().c_str());
                        }
                    }
                }

                LOGEXIT(1, "Composite::resolveWires");
            }

            void Composite::addWSDLDefinition(commonj::sdo::DataObjectPtr wsdlModel)
            {
                LOGENTRY(1, "Composite::addWSDLDefinition");
                WSDLDefinition* wsdlDefinition = new WSDLDefinition(wsdlModel);
                wsdlDefinitions[wsdlDefinition->getNamespace()] = wsdlDefinition;
                LOGEXIT(1, "Composite::addWSDLDefinition");

            }

            WSDLDefinition* Composite::findWSDLDefinition(const std::string& wsdlNamespace )
            {
                return wsdlDefinitions[wsdlNamespace];

            }

            commonj::sdo::XSDHelperPtr Composite::getXSDHelper()
            {
                if (xsdHelper == 0)
                {
                    xsdHelper = commonj::sdo::HelperProvider::getXSDHelper();
                }

                return xsdHelper;
            }

            commonj::sdo::XMLHelperPtr Composite::getXMLHelper()
            {
                if (xmlHelper == 0)
                {
                    xmlHelper = commonj::sdo::HelperProvider::getXMLHelper(getXSDHelper()->getDataFactory());
                }

                return xmlHelper;
            }

            commonj::sdo::DataFactoryPtr Composite::getDataFactory()
            {
                return getXSDHelper()->getDataFactory();
            }
            
        } // End namespace model
    } // End namespace sca
} // End namespace tuscany

File: modules/wsgi/composite.py
#!/usr/bin/python
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#  
#    http://www.apache.org/licenses/LICENSE-2.0
#    
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

# Composite deployment and integration with WSGI

from wsgiref.simple_server import make_server
from wsgiref.handlers import CGIHandler
from wsgiref.util import request_uri
from wsgiref.util import FileWrapper
from os import environ
import os.path
import hashlib
from sys import stderr, argv
from util import *
from scdl import *
from atomutil import *
from jsonutil import *

# Cache the deployed components between requests
comps = None

# Return the path of an HTTP request
def requestPath(e):
    return e.get("PATH_INFO", "")

# Return the method of an HTTP request
def requestMethod(e):
    return e.get("REQUEST_METHOD", "")

# Return the method of an HTTP request
def requestContentType(e):
    return e.get("CONTENT_TYPE", "")

# Return the request body input stream
def requestBody(e):
    i = e.get("wsgi.input", None)
    if i == None:
        return ()
    l = int(e.get("CONTENT_LENGTH", "0"))
    return (i.read(l),)

def requestIfNoneMatch(e):
    return e.get("HTTP_IF_NONE_MATCH", "");

# Hash a list of strings into an MD5 signature
def md5update(md, s):
    if isNil(s):
        return md.hexdigest()
    md.update(car(s))
    return md5update(md, cdr(s))

def md5(s):
    return md5update(hashlib.md5(), s)

# Return an HTTP success result
def result(e, r, st, h = (), b = None):
    if st == 201:
        r("201 Created", list(h))
        return ()

    if st == 200:
        if b == None:
            r("200 OK", list(h))
            return ()

        # Handle etags to minimize bandwidth usage
        md = md5(b)
        if (md == requestIfNoneMatch(e)):
            r("304 Not Modified", list((("Etag", md), ("Expires", "Tue, 01 Jan 1980 00:00:00 GMT"))))
            return ()
        r("200 OK", list(h + (("Etag", md), ("Expires", "Tue, 01 Jan 1980 00:00:00 GMT"))))
        return b

    if st == 301:
        r("301 Moved Permanently", list(h))

    return failure(e, r, 500)

# Return an HTTP failure result
def failure(e, r, st):
    s = "404 Not Found" if st == 404 else str(st) + " " + "Internal Server Error"
    r(s, list((("Content-type", "text/html"),)))
    return ("<html><head><title>"+ s + "</title></head><body><h1>" + s[4:] + "</h1></body></html>",)

# Return a static file
def fileresult(e, r, ct, f):

    # Read the file, return a 404 if not found
    p = "htdocs" + f
    if not os.path.exists(p):
        return failure(e, r, 404)
    c = tuple(FileWrapper(open("htdocs" + f)))

    # Handle etags to minimize bandwidth usage
    md = md5(c)
    r("200 OK", list((("Content-type", ct),("Etag", md))))
    return c

# Converts the args received in a POST to a list of key value pairs
def postArgs(a):
    if isNil(a):
        return ((),)
    l = car(a);
    return cons(l, postArgs(cdr(a)))

# Return the URL used to sign out
def signout(ruri):
    try:
        from google.appengine.api import users
        return users.create_logout_url(ruri)
    except:
        return None

# Return the URL used to sign in
def signin(ruri):
    try:
        from google.appengine.api import users
        return users.create_login_url(ruri)
    except:
        return None

# WSGI application function
def application(e, r):
    m = requestMethod(e)
    fpath = requestPath(e)

    # Serve static files
    if m == "GET":
        if fpath.endswith(".html"):
            return fileresult(e, r, "text/html", fpath)
        if fpath.endswith(".js"):
            return fileresult(e, r, "application/x-javascript", fpath)
        if fpath.endswith(".png"):
            return fileresult(e, r, "image/png", fpath)
        if fpath == "/":
            return result(e, r, 301, (("Location", "/index.html"),))

        # Debug hook
        if fpath == "/debug":
            return result(e, r, 200, (("Content-type", "text/plain"),), ("Debug",))

        # Sign in and out
        if fpath == "/login":
            redir = signin("/")
            if redir:
                return result(e, r, 301, (("Location", redir),))
        if fpath == "/logout":
            redir = signout(signin("/"))
            if redir:
                return result(e, r, 301, (("Location", redir),))

    # Find the requested component
    path = tokens(fpath)
    uc = uriToComponent(path, comps)
    uri = car(uc)
    if uri == None:
        return failure(e, r, 404)
    comp = cadr(uc)

    # Call the requested component function
    id = path[len(uri):]
    if m == "GET":
        v = comp("get", id)
        
        # Write returned content-type / content pair
        if not isinstance(cadr(v), basestring):
            return result(e, r, 200, (("Content-type", car(v)),), cadr(v))
        
        # Write an ATOM feed or entry
        if isNil(id):
            return result(e, r, 200, (("Content-type", "application/atom+xml;type=feed"),), writeATOMFeed(feedValuesToElements(v)))
        return result(e, r, 200, (("Content-type", "application/atom+xml;type=entry"),), writeATOMEntry(entryValuesToElements(v)))

    if m == "POST":
        ct = requestContentType(e)

        # Handle a JSON-RPC function call
        if ct.find("application/json-rpc") != -1 or ct.find("text/plain") != -1:
            json = elementsToValues(readJSON(requestBody(e)))
            args = postArgs(json)
            jid = cadr(assoc("'id", args))
            func = funcName(cadr(assoc("'method", args)))
            params = cadr(assoc("'params", args))
            v = comp(func, *params)
            return result(e, r, 200, (("Content-type", "application/json-rpc"),), jsonResult(jid, v))

        # Handle an ATOM entry POST
        if ct.find("application/atom+xml") != -1:
            ae = entryValue(readATOMEntry(requestBody(e)))
            v = comp("post", id, ae)
            if isNil(v):
                return failure(e, r, 500)
            return result(e, r, 201, (("Location", request_uri(e) + "/" + "/".join(v)),))
        return failure(e, r, 500)
    
    if m == "PUT":
        # Handle an ATOM entry PUT
        ae = entryValue(readATOMEntry(requestBody(e)))
        v = comp("put", id, ae)
        if v == False:
            return failure(e, r, 404)
        return result(e, r, 200)
    
    if m == "DELETE":
        v = comp("delete", id)
        if v == False:
            return failure(e, r, 404)
        return result(e, r, 200)
    
    return failure(e, r, 500)

# Return the WSGI server type
def serverType(e):
    return e.get("SERVER_SOFTWARE", "")

def main():
    # Read the deployed composite and evaluate the configured components
    global comps
    if comps == None:
        domain = "domain.composite" if os.path.exists("domain.composite") else "domain-test.composite"
        comps = evalComponents(components(parse(domain)))

    # Handle the WSGI request with the WSGI runtime
    st = serverType(environ)
    if st.find("App Engine") != -1 or st.find("Development") != -1:
        from google.appengine.ext.webapp.util import run_wsgi_app
        run_wsgi_app(application)
    elif st != "":
        CGIHandler().run(application)
    else:
        make_server("", int(argv[1]), application).serve_forever()

# Run the WSGI application
if __name__ == "__main__":
    main()

